moped-editor/src/auth/user.js

```diff
@@ -0,0 +1,100 @@
+export const UserContext = React.createContext(null);
+
+// Create a "controller" component that will calculate all the data that we need to give to our
+// components bellow via the `UserContext.Provider` component. This is where the Amplify will be
+// mapped to a different interface, the one that we are going to expose to the rest of the app.
+export const UserProvider = ({ children }) => {
+  const [user, setUser] = React.useState(null);
+
+  React.useEffect(() => {
+    // Configure the keys needed for the Auth module. Essentially this is
+    // like calling `Amplify.configure` but only for `Auth`.
+    /**
+     * AWS Amplify
+     * @see https://github.com/aws-amplify/amplify-js
+     */
+
+    Amplify.Logger.LOG_LEVEL = "DEBUG";
+
+    Amplify.configure({
+      Auth: {
+        mandatorySignIn: true,
+        region: config.cognito.REGION,
+        userPoolId: config.cognito.USER_POOL_ID,
+        identityPoolId: config.cognito.IDENTITY_POOL_ID,
+        userPoolWebClientId: config.cognito.APP_CLIENT_ID
+      },
+      API: {
+        endpoints: [
+          {
+            name: "testApi",
+            endpoint: config.apiGateway.URL,
+            region: config.apiGateway.REGION
+          }
+        ]
+      }
+    });
+    // attempt to fetch the info of the user that was already logged in
+    Auth.currentAuthenticatedUser()
+      .then(user => setUser(user))
+      .catch(() => setUser(null));
+  }, []);
+
+  // We make sure to handle the user update here, but return the resolve value in order for our components to be
+  // able to chain additional `.then()` logic. Additionally, we `.catch` the error and "enhance it" by providing
+  // a message that our React components can use.
+  const login = (usernameOrEmail, password) =>
+    Auth.signIn(usernameOrEmail, password)
+      .then(cognitoUser => {
+        setUser(cognitoUser);
+        return cognitoUser;
+      })
+      .catch(err => {
+        if (err.code === "UserNotFoundException") {
+          err.message = "Invalid username or password";
+        }
+
+        // ... (other checks)
+
+        throw err;
+      });
+
+  // same thing here
+  const logout = () =>
+    Auth.signOut().then(data => {
+      setUser(null);
+      return data;
+    });
+
+  // Make sure to not force a re-render on the components that are reading these values,
+  // unless the `user` value has changed. This is an optimisation that is mostly needed in cases
+  // where the parent of the current component re-renders and thus the current component is forced
+  // to re-render as well. If it does, we want to make sure to give the `UserContext.Provider` the
+  // same value as long as the user data is the same. If you have multiple other "controller"
+  // components or Providers above this component, then this will be a performance booster.
+  const values = React.useMemo(() => ({ user, login, logout }), [user]);
```

Comment:
This is cool! Glad to see a solution to the problem that we all discussed on Monday about potential performance cost. Might be cool add to VZE one day too.